src/compile/data/optimizers.ts

```diff
@@ -11,34 +11,63 @@ import {TimeUnitNode} from './timeunit';
 import {CalculateNode, FilterNode} from './transforms';
 
 /**
- * Start optimization path at the leaves. Useful for merging up things.
+ * Start optimization path at the leaves. Useful for merging up or removing things.
+ *
+ * If the callback returns true, the recursion continues.
  */
-export function optimizeFromLeaves(f: (node: DataFlowNode) => void) {
+export function iterateFromLeaves(f: (node: DataFlowNode) => boolean) {
   function optimizeNextFromLeaves(node: DataFlowNode) {
-    if (node.parent instanceof SourceNode) {
+    if (node instanceof SourceNode) {
       return;
-    } else if (!node || !node.parent) {
-      throw new Error('A source node cannot have parents and roots haev to be source nodes.');
     }
 
     const next = node.parent;
-    f(node);
-    optimizeNextFromLeaves(next);
+    if (f(node)) {
+      optimizeNextFromLeaves(next);
+    }
   }
 
   return optimizeNextFromLeaves;
 }
 
-
-export function parse(node: DataFlowNode) {
+/**
+ * Move parse nodes up to forks.
+ */
+export function moveParseUp(node: DataFlowNode) {
   const parent = node.parent;
 
   // move parse up by merging or swapping
   if (node instanceof ParseNode) {
+    if (parent instanceof SourceNode) {
+      return false;
+    }
+
+    if (parent.numChildren() > 1) {
+      return true;
```

Comment:
I guess this return true because it will continue move other parse upstream up?  If so, maybe add comment?